mitodl/mit-open
# Use DRF NamespaceVersioning to manage OpenAPI api versions

### What are the relevant tickets?
Closes #402 

### Description (What does it do?)
This PR introduces DRF's [NamespaceVersioning](https://www.django-rest-framework.org/api-guide/versioning/#namespaceversioning) class to version our APIs. 

With the current settings in this PR:
- `drf-spectacular` now generates documentation from our main `open_discussions.urls` url configuration; two separate versions, `openapi/specs/v0.yml` and `openapi/specs/v1.yml`, are generated.
    - previously we maintained a separate list `openapi/v1/urls_spectaular.yaml` for the schema/docs, and would have had to maintain one such list per version
- in order to appear in our generated OpenAPI specifications, API endpoints **must** have a version provided by a url [namespace](https://docs.djangoproject.com/en/5.0/topics/http/urls/#url-namespaces)
- the version must be either `"v0"` or `"v1"` (we can add more, via `settings.REST_FRAMEWORK["ALLOWED_VERSIONS"]`.
    - I made learning resource + search endpoints v1, and other things that are somewhat experimental/to-be-developed-more (ckeditor, channels) v0. There's one exception (articles) with a note.


### How can this be tested?
1. Visit the rendered docs. (Check `swagger-ui`, too, if you want.)
    - The v0 docs: http://localhost:8063/api/v0/schema/redoc/
    - The v1 docs: http://localhost:8063/api/v1/schema/redoc/
2. Manually edit the schemas so they are out of sync:
    - Edit `openapi/schemas/v0.yaml` and/or `openapi/schemas/v1.yaml` ... Just change a description or something.
    - Run `docker compose exec web ./scripts/test/openapi_spec_check.sh` it should exit with a nonzero status code.
3. Regenerate the docs:
    - Run (outside docker) `./scripts/generate_openapi.sh`
    - Check `git status`; the v0 and v1 yaml files should now be un-modified (changes from previous step overriden).
4. You could try changing an endpoint's version or adding a new endpoint. For example, change the `learning_resources_search/urls.py` urlpatterms to

    ```python
    urlpatterns = [
        re_path(r"^api/v0/", include((v1_urls, "v0"))),
        re_path(r"^api/v1/", include((v1_urls, "v1"))),
    ]
    ```

     and now the search endpoints should be available under both `api/v0` and `api/v1` and appear in both documentations.


### Additional Context

**Request Versions:** With NamespaceVersioning, the version is also available on the request object via `request.version`. This can be used (for example) to swap serializer classes based on version. I don't necessarily expect us to use this functionality, though we could. Right now the request versioning is just for schema + doc generation.
